Check pull request with astyle for proper code style

tstibor opened the pull request:
Leverage astyle source code indenter and formatter for pre-checking pull request code style.
If code style is in agreement with settings listed in .astylerc, continue with additional pull request handling, otherwise
terminate with warning of style warning.

etracer65 commented:
I don't think we want to force anybody submitting a PR to be responsible for fixing all code styling problems in the files affected by their PR.

tstibor commented:
You are right, this will affect the whole file and not just the diff. How about changing it to something like that:

```
for f in "${FILE_LIST[@]}";
	astyle --options=.astylerc < "${f}" | cmp -s "${f}" -
	if [ $? -ne 0 ]; then
		echo "Pull request changing file ${f} does not respect the agreed coding style." >&2
	        exit 1
	fi
done
```

tstibor commented:
Yes you are right, it reports all code styling problems of the changed file rather than the diff styling problems. The most straightforward way to solve this issue, is by using a git pre-commit hook, but this needs to be done by hand (e.g. $ln -s <...>) if I am not mistaken,

etracer65 commented:
I'm not sure how it will handle potentially small snippets that by themselves appear to violate the coding style but in the context of the surrounding code are correct. Like:
```
     if (something) {
         do_something;
         do_something;
+    } else {
+        do_something_else;
     }
```
The changed part of the `diff` doesn't have any context by itself to show that the overall code is correctly styled.

The key point is that while we want people to follow the stylistic guidelines we don't want to have them constantly bumping into hard checks that frustrate and discourage contribution. Previous attempts to enforce style programmatically have failed for these reasons.

mikeller commented:
@tstibor: I can see that there is value in what you are trying to do - it would be nice to have something that allowed tracked coding style in a similar manner to what SonarCloud does, and only alerted for newly introduced infarctions. SonarQube itself is unfortunately not usable for the firmware, as it will report 1000s of infarctions based on conditionals.
But any proposal for a way to get this behaviour is appreciated.

tstibor commented:
Ok, I will revival the issue and investigate what's the best approach to style check PR's

tstibor commented:
Hi there,

the following codestyle checking script seems to work well on git diff snippets:

```
#!/bin/sh

TMP_FILE_DIFF=$(mktemp /tmp/betaflight.diff.XXXXXXXXX)
TMP_FILE_ASTYLE=$(mktemp /tmp/betaflight.astyle.XXXXXXXXX)
TMP_FILE_WARNINGS=$(mktemp /tmp/betaflight.warning.XXXXXXXXX)

git diff --function-context --unified=1 | sed -e '/diff --git/,+3d;/^@@/d;/^-/d;s/^+/ /' | cut -d' ' -f2- > ${TMP_FILE_DIFF}
astyle -n --options=.astylerc < ${TMP_FILE_DIFF} > ${TMP_FILE_ASTYLE}
diff -u ${TMP_FILE_DIFF} ${TMP_FILE_ASTYLE} | tail -n +3 | sed -e '/^-/d;/^@@/d;s/^+/[STYLE WARNING] /' > ${TMP_FILE_WARNINGS}

if [ -s ${TMP_FILE_WARNINGS} ] ; then
    echo "Pull request has style warning(s), for more details please see"
    echo "https://github.com/betaflight/betaflight/blob/master/docs/development/CodingStyle.md\n"
    cat ${TMP_FILE_WARNINGS}
fi

rm -f ${TMP_FILE_DIFF} ${TMP_FILE_ASTYLE} ${TMP_FILE_WARNINGS}
```
I added for testing extra spaces, tabs ...
```
$ git diff --function-context --unified=1
diff --git a/src/main/cli/cli.c b/src/main/cli/cli.c
index 81026071a..a90c627b6 100644
--- a/src/main/cli/cli.c
+++ b/src/main/cli/cli.c
@@ -380,9 +380,9 @@ void cliPrintLine(const char *str)
 #ifdef MINIMAL_CLI
 #define cliPrintHashLine(str)
 #else
 static void cliPrintHashLine(const char *str)
 {
-    cliPrint("\r\n# ");
+      cliPrint("\r\n# ");
     cliPrintLine(str);
 }
 #endif
@@ -390,4 +390,5 @@ static void cliPrintHashLine(const char *str)
 static void cliPutp(void *p, char ch)
 {
     bufWriterAppend(p, ch);
+    int c;
 }
@@ -416,7 +417,7 @@ static bool cliDumpPrintLinef(dumpFlags_t dumpMask, bool equalsDefault, const ch
 
-static void cliWrite(uint8_t ch)
+ static void cliWrite(uint8_t ch)
 {
     if (cliWriter) {
         bufWriterAppend(cliWriter, ch);
     }
 }
@@ -424,15 +425,15 @@ static void cliWrite(uint8_t ch)
 static bool cliDefaultPrintLinef(dumpFlags_t dumpMask, bool equalsDefault, const char *format, ...)
 {
     if ((dumpMask & SHOW_DEFAULTS) && !equalsDefault) {
         cliWrite('#');
 
         va_list va;
         va_start(va, format);
         cliPrintfva(format, va);
         va_end(va);
         cliPrintLinefeed();
         return true;
     } else {
-        return false;
+               return false;
     }
 }
diff --git a/src/main/main.c b/src/main/main.c
index 2d93e83af..ea237adba 100644
--- a/src/main/main.c
+++ b/src/main/main.c
@@ -41,10 +41,10 @@ int main(void)
 void FAST_CODE FAST_CODE_NOINLINE run(void)
 {
     while (true) {
         scheduler();
         processLoopback();
 #ifdef SIMULATOR_BUILD
-        delayMicroseconds_real(50); // max rate 20kHz
+            delayMicroseconds_real(50); // max rate 20kHz
 #endif
     }
 }
```
The output of codestyle script looks as follows:
```
$ ./codestyle.sh                                                                                                                                                                                                                                           
Pull request has style warning(s), for more details please see
https://github.com/betaflight/betaflight/blob/master/docs/development/CodingStyle.md

 #else
 static void cliPrintHashLine(const char *str)
 {
[STYLE WARNING]     cliPrint("\r\n# ");
     cliPrintLine(str);
 }
 #endif
     int c;
 }
 
[STYLE WARNING] static void cliWrite(uint8_t ch)
 {
     if (cliWriter) {
         bufWriterAppend(cliWriter, ch);
         cliPrintLinefeed();
         return true;
     } else {
[STYLE WARNING]         return false;
     }
 }
 void FAST_CODE FAST_CODE_NOINLINE run(void)
         scheduler();
         processLoopback();
 #ifdef SIMULATOR_BUILD
[STYLE WARNING]         delayMicroseconds_real(50); // max rate 20kHz
 #endif
     }
 }
```
Any comments?

Cheers
 Thomas

tstibor commented:
@mikeller , @etracer65 : I created a separate BASH script check_cstyle.sh for checking the code style.

From TRAVIS, the PR modified files FILE_LIST=$(git diff --name-only $COMMIT_RANGE) are used as script input files.
The scripts checks whether file exists and is C oder header file. With
git diff --function-context --unified=1 HEAD ${FILE_TO_CHECK} one gets "complete" function diffs. The subsequent "sed" and "cut" removes the diff typical output, so one gets kind of clean C code. This is feed into astyle and the output diff is compared.
If there is a diff, I replace the "+ diff" with [STYLE WARNING] and also outputs at the end:

echo -e "File ${FILE_TO_CHECK} has style warning(s), see" \
         "https://github.com/betaflight/betaflight/blob/master/docs/development/CodingStyle.md"

mikeller commented:
@tstibor: I think having this as a separate script is a good approach - we can always add it as a separate, non-required build job in Azure - this way it will be run, and it will provide a warning for the failed build if it fails, but it will not stop mergeing.
One thing though is that we are not using Travis any more, as they have changed ownership, and have now essentially stopped providing their service for free to open source projects - so we will need to find a different way to identify the qualifying files.

tstibor commented:
@mikeller: Added lines to azure-pipelines.yml
```
- bash: for f in `git diff --name-only`; do src/utils/check_cstyle.sh "${f}"; done
  displayName: 'C style check'
```
It seems to work:
```
Starting: C style check
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.179.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
Generating script.
Script contents:
for f in `git diff --name-only`; do src/utils/check_cstyle.sh "${f}"; done
========================== Starting Command Output ===========================
/bin/bash --noprofile --norc /home/vsts/work/_temp/9d0a0bd0-0451-4e79-8416-93019dc5cf97.sh
Finishing: C style check
```
I am however, not sure whether the `cat ${TMP_FILE_WARNINGS}` in src/utils/check_cstyle.sh is propagated, so the user gets
this information e.g. by mail. Probably `cat `to stderr is here the way to go. In addition how are return codes handled from calling a bash script within azure-pipelines.yml.
Script src/utils/check_cstyle.sh exit with 1 when warning are found, otherwise exit with 0.

mikeller commented on azure-pipelines.yml:
```diff
@@ -42,8 +42,10 @@ stages:
-    - script: sudo apt-get install -y libblocksruntime-dev
+    - script: sudo apt-get install -y libblocksruntime-dev astyle
       displayName: 'Install extra packages'
+    - bash: for f in `git diff --name-only`; do src/utils/check_cstyle.sh "${f}"; done
```
Not sure this will work - since the repository is checked out into the branch at this point, the `git diff --name-only` will return an empty list.
A better way will be `git diff --name-only master`, but this has shortcomings as well:
- spurious files if there were commits to `master` after this branch was branched;
- a more or less arbitrary list of files if the pull request is against a different branch than `master`;
- obviously this will not work for CI builds against `master`, but I think this case can be ignored (or better: running of style checks can be suppressed in this case).

tstibor replied:
How about: git diff --name-only "$SYSTEM_PULLREQUEST_SOURCEBRANCH", according to
[Migrate from Travis to Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/migrate/from-travis?view=azure-devops), that is "The name of the branch where the pull request originated."

Regarding separate job, that is very good. I will put this in a separate job, e.g.:

```
 - job: 'Style check'
    pool:
      vmImage: 'ubuntu-16.04'
    steps:
    - script: sudo apt-get install -y astyle
      displayName: 'Install extra packages'
    - bash: for f in `git diff --name-only "$SYSTEM_PULLREQUEST_SOURCEBRANCH"`; do src/utils/check_cstyle.sh "${f}"; done
      displayName: 'C style check'
```

mikeller replied:
@tstibor: Sorry, only picked this up again now.
Was there a problem with the above approach? The current version of `git diff HEAD HEAD~ --name-only` has got the issue that it will miss some files if the pull request contains more than one commit.

mikeller commented:
@tstibor: I think re notification, a good way to go on about this will be to do the style checks in a separate job (see `job` stanzas in `azure.yml`) - this way it will appear as a separate item that can succeed or fail in the GitHub 'Checks' tab:

![image](https://user-images.githubusercontent.com/4742747/107507871-10f41500-6c05-11eb-87bc-df26c5c59ac4.png)

It is possible to make only certain jobs required to pass for a successful build, so we can implement the 'warning only' behaviour by having the 'style check' job fail, but make it not required to pass.

tstibor commented:
Unfortunately, the command: `git diff --name-only master` results in the error:
```
for f in `git diff --name-only master`; do src/utils/check_cstyle.sh "${f}"; done
========================== Starting Command Output ===========================
/bin/bash --noprofile --norc /home/vsts/work/_temp/547d42a6-149d-4794-8743-76c2083b597d.sh
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Finishing: C style check
```

mikeller commented:
@tstibor: That's a bit of a bummer. And this also applies when you use it as `git diff --name-only origin/master`, or even better, as `git diff --name-only ${SYSTEM_PULLREQUEST_SOURCEBRANCH}`?

tstibor commented:
@mikeller:
Tried `git diff --name-only ${SYSTEM_PULLREQUEST_SOURCEBRANCH}`, it resulted in 
```
for f in `git diff --name-only ${SYSTEM_PULLREQUEST_SOURCEBRANCH}`; do src/utils/check_cstyle.sh "${f}"; done
========================== Starting Command Output ===========================
/bin/bash --noprofile --norc /home/vsts/work/_temp/2bd0902d-236f-48ae-96cf-35d2e3549835.sh
fatal: ambiguous argument 'pre_check_astyle': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Finishing: C style check
```
However, `git diff --name-only origin/master` works :+1: 
```
Starting: C style check
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.182.0
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
Generating script.
Script contents:
for f in `git diff --name-only origin/master`; do src/utils/check_cstyle.sh "${f}"; done
========================== Starting Command Output ===========================
/bin/bash --noprofile --norc /home/vsts/work/_temp/a9c907ac-d2aa-4401-b051-911fb5986b21.sh
Finishing: C style check
```
The latest update of this PR is with: `git diff --name-only origin/master`
There is still the drawback of master branch only,  but at least on master it seems to works.

mikeller commented:
@tstibor: Good going, we are getting closer. I think the solution we have got now is closing in on where we want to be for pull requests. With respect to making this work on changes in `master` as well, after looking into the Azure CI variables, the solution will probably mean to do an 'if pull request then ... else' - not sure if there is value in this.

Re what you found, looks like we have to supply an `origin/` prefix to force Azure to compare against the remote version - makes sense.

One more thing we can try, in order to eliminate any change to `master` after the pull request branch was cut, will be:

```
PULL_REQUEST_BRANCH_ROOT=git merge-base origin/${SYSTEM_PULLREQUEST_SOURCEBRANCH} origin/master
for f in `git diff --name-only ${PULL_REQUEST_BRANCH_ROOT}`; do src/utils/check_cstyle.sh "${f}"; done
```

haslinghuis commented:
@tstibor Any update on this, this is great work?

tstibor commented:
> @tstibor Any update on this, this is great work?

I rebased this PR and modified `vmImage: 'ubuntu-16.04'` to `vmImage: '$(vmImage)'`
C style checking PR's for origin/master works, but not the other branches.
